### PR TITLE
Introduce Arguments / Module class

### PR DESCRIPTION
**Pull request by nupplaphil**
Redux of #7496 because of growing complexity

The Arguments and Module class are now separating code of the `App` class into own classes. They are now better to test and I focused the used logic, which was currently placed all over the `App` class (like `query_string`, `isBackend`, ...).

With reducing the code of `App`, we also reduce the so called `cyclomatic complexity`, which is a good hint of maintainable code :-).

And we can now use arguments of a call without loading the whole app for it, which is good for tests and for backend calls as well :-)

Checks:
- [x] added tests
- [x] local node works
- [x] friendica.philipp.info works (with error logging on)
- [x] installer works (I fixed a bug ..)

FollowUp (which was the reason to break the changes into little pieces):
- Extract frontend logic of `App` into a own class `Frontend`
- Refactor `L10n` to be usable without loading the whole app and therefore be testable and it can be used for console-output without loading user defined configs (reduced in #7496 )
- Refactor `Session` to be testable (is a prerequisite for #7496 )

### CODE REVIEW

File: index.php [hunk 1]
```diff
@@ -18,4 +18,4 @@
 
 $a = \Friendica\BaseObject::getApp();
 
-$a->runFrontend();
+$a->runFrontend($dice->create(\Friendica\App\Module::class), $dice->create(\Friendica\App\Router::class), $dice->create(\Friendica\Core\Config\PConfiguration::class));
```
Comment (MrPetovan) [inline index.php]: Please add new lines between each argument.

File: src/App/Arguments.php [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+	/**
+	 * Determine the arguments of the current call
+	 *
+	 * @param array $server The $_SERVER variable
+	 * @param array $get    The $_GET variable
+	 *
+	 * @return Arguments The determined arguments
+	 */
+	public function determine(array $server, array $get)
```
Comment (MrPetovan) [inline src/App/Arguments.php]: Shouldn't this be in the constructor? It made sense for static classes to have `init` or `determine` methods but do we still need them with class instances?

File: src/App/Mode.php [hunk 3]
```diff
@@ -54,50 +36,46 @@ public function __construct(BasePath $basepath, Database $database, ConfigCache
...
 	 *
-	 * @param string $basePath the Basepath of the Application
-	 *
-	 * @return Mode returns itself
+	 * @return Mode returns the determined mode
 	 *
 	 * @throws \Exception
 	 */
-	public function determine($basePath = null)
+	public function determine(BasePath $basepath, Database $database, ConfigCache $configCache)
```
Comment (MrPetovan) [inline src/App/Mode.php]: Same question about constructor and `determine`/`init` methods.

File: src/App/Module.php [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+	 *
+	 * @param Arguments                 $args   The Friendica execution arguments
+	 * @param Router                    $router The Friendica routing instance
+	 * @param Core\Config\Configuration $config The Friendica Configuration
+	 *
+	 * @return Module The determined module of this call
+	 *
+	 * @throws \Friendica\Network\HTTPException\InternalServerErrorException
+	 */
+	public function determineClass(Arguments $args, Router $router, Core\Config\Configuration $config)
```
Comment (MrPetovan) [inline src/App/Module.php]: Why is the method returning a different `App\Module` object?

**Comment by nupplaphil** [comment]
I was a little bit investigating how to improve even more the code and I read through some articles about
- *Immutablity* (https://www.yegor256.com/2014/06/09/objects-should-be-immutable.html)
- *Lean constructor*

The reason for Arguments, Module and Mode to have no constructor business logic is that we can now build them without any additional, implicit loading inside the constructor. Therefore I'm not even have to mock these objects anymore. For example if I do want to test a function in case that the `App\Mode` is in maintenance mode, I do have to create `new App\Mode(14)` and that's it :-). 

Another reason is that we now don't have to store the whole loading structures (Database,  Configuration, ..) inside the class, because they are just used in case of auto-"determined". So we can define when we want to load the environment.

Returning a "new" object after the determination is based on the "immutable" pattern, which says that changing the state of an object should result in a new object. This should enable coders to be sure which state the current class has in any circumstances. It's not practicable for every class, but in case of these classes, we just have the state after loading and this shouldn't be altered.

Another point is that immutable classes are per se multi-thread safe because no thread is possible to alter the state of the class, which would harm a parallel run.

I don't want to enforce these practices, but they seems right for me. For example I'd like to have a `Module` class injected without the `determineClass` method, so it's possible to use it without atuomatically loading the whole Database instance. Another example will become the `L10n` class, so we can choose use it with the system language, the user language and optionally with all loaded addons. The first one is the one, which I do want to use in the SystemCheck, because so I can create a most-lean call without any other, implicit calls

**Comment by MrPetovan** [comment]
Thanks for the elaboration!

**Comment by MrPetovan** [comment]
Another removed method fatal error:
```
PHP Fatal error:  Uncaught Error: Call to undefined method Friendica\App::isBackend() in src/Core/Worker.php:1129
Stack trace:
#0 src/Model/Contact.php(1405): Friendica\Core\Worker::add(40, 'UpdateContact', 109247, 'force')
#1 include/conversation.php(905): Friendica\Model\Contact::getIdForURL('https://mastodo...', 0, true)
#2 src/Object/Post.php(390): item_photo_menu(Array)
#3 src/Object/Post.php(454): Friendica\Object\Post->getTemplateData(Array, 2)
#4 src/Object/Thread.php(188): Friendica\Object\Post->getTemplateData(Array)
#5 include/conversation.php(770): Friendica\Object\Thread->getTemplateData(Array)
#6 mod/network.php(344): conversation(Object(Friendica\App), Array, Object(Friendica\Content\Pager), 'network', 0, false, '`received`', 1)
 in src/Core/Worker.php on line 1129
```
I'm reverting this PR.

**Pull request merged**